src/server/rot.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright lowRISC contributors.
+// Copyright (c) 2021 Western Digital
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! A `manticore` "server" AC and PR RoTs.
+//!
+//! This module provides structures for serving responses to a host making
+//! requests to a PA-RoT and AC-RoT.
+
+use crate::crypto::rsa;
+use crate::hardware;
+use crate::mem::Arena;
+use crate::net;
+use crate::protocol;
+use crate::server::options::Options;
+use crate::server::response::Response;
+use crate::server::Error;
+
+use crate::server::handler::prelude::*;
+
+/// A RoT, or "Root of Trust", server.
+///
+/// This type implements the request -> response "business logic" of the
+/// host <-> PA-RoT interaction or the PA-RoT <-> AC-RoT. That is, it accepts
+/// input and output buffers, and from those, parses incoming requests and
+/// processes them into responses.
+///
+/// This is a generic RoT type that can be used as either a AC-RoT or a PA-RoT.
+pub(crate) struct Rot<'a, Identity, Reset, Rsa> {
```

Comment:
This is probably a useful abstraction, but the PA and AC RoTs are both a little different. What's your plan for exposing this configurability concretely? (Some Rust pseudocode would be useful for illustration.)

Reply:
This should be pretty straightforward to handle.

For things that only a PA/AC RoT can do we just add a function to `AcRot` or `PaRot` and not `Rot`. For example `PaRot` should have a `start_request()` function to initiate requests while the `AcRot` won't.

For different logic in shared functionality we can either handle it in the `AcRot`/`PaRot` specific function before calling the generic `Rot` one:

```rust
impl<'a, Identity, Reset, Rsa> Respond for PaRot<'a, Identity, Reset, Rsa>
where
    Identity: hardware::Identity,
    Reset: hardware::Reset,
    Rsa: rsa::Builder,
{
    fn process_request<'req>(
        &mut self,
        host_port: &mut dyn net::HostPort,
        arena: &'req impl Arena,
    ) -> Result<(), Error> {
        if some_rot_condition {
            return self.do_something();
        }

        self.rot.process_request(host_port, arena)
    }
}
```

Or the generic `Rot` can use the `opts.networking` to take a different action based on PA/AC RoT